mLRS/CommonTx/jr_pin5_interface.h

```diff
@@ -133,6 +133,15 @@ void tPin5BridgeBase::Init(void)
     LL_USART_Enable(UART_UARTx);
     gpio_init_af(UART_RX_IO, IO_MODE_INPUT_PD, UART_IO_AF, IO_SPEED_VERYFAST);
 #endif
+// swaps the Tx and Rx pins for E77 easysolder as Tx
+#if defined JRPIN5_RX_TX_INVERT_SWAP_INTERNAL
+    LL_USART_Disable(UART_UARTx);
+    LL_USART_SetTXPinLevel(UART_UARTx, LL_USART_TXPIN_LEVEL_INVERTED);
+    LL_USART_SetRXPinLevel(UART_UARTx, LL_USART_RXPIN_LEVEL_INVERTED);
+    LL_USART_SetTXRXSwap(UART_UARTx, LL_USART_TXRX_SWAPPED);
+    LL_USART_Enable(UART_UARTx);
+    gpio_init_af(UART_RX_IO, IO_MODE_INPUT_PD, UART_IO_AF, IO_SPEED_VERYFAST);
```

Comment:
I need to object here. Such a config of the gpios is electrically just not sound, which implies that if it "works" it would be susceptible to all sorts of distortions, variations, noise. Also, this test doesn't prove that every and each messages are just transmitted well. It's a bit along the same lines of finding that a resistors "works", the voltage levels may happen to just cut the edges dependent on what other resistances are around, but that's a delicate balance, introduce device dependency as well as giving noise much impact. Maybe you have an oscilloscope to trace the signal, to see how well/badly it swings between 0V and 3.3V.
I have to say I'm actually quite puzzled you find it to work to some extend at all, since the tx pin is now configured as input and not output.
Anyway, my point, we want to do it electrically proper and not in a wonky solala fashion :)

I think I haven't ever heard that the gpio config shouldn't work anymore once the peripheral is configured. I guess we need to check with the datasheet. If that should be indeed an issue it should be easy to resolve by simply moving the gpio init to before the enable of the peripheral.

EDIT: I think I actually have experimental "proof" that the gpio init works after enable. I do use the uart_tx_enablepin() function in the uart library in quite some projects of mine, and these would not work if the functuion would not do what it is supposed to do.

Reply:
That's all fair enough :)  I spent some time and wanted to share 2 notes:

1. Checked with and without that line on the scope and both signals looked clean to me (using the 1K resistor).
2. As both the Tx and Rx are initialized with alternate function modes, this discussion seems to show that the peripheral controls whether they are input / output https://community.st.com/t5/stm32-mcu-products/alternate-function-input-on-stm32f4/m-p/540378

In addition, I had the CubeIDE configure a UART and it initialized both pins as output push-pull...

![image](https://github.com/olliw42/mLRS/assets/41841496/4384bc5c-8372-4276-a768-701aa100de2b)

So while it does seem to work, I certainly can't say 100% why.

Reply:
as much as I understand it, af overwrites if a pin is input or output, but only this. (I think that's what you are saying too)

Note that the purpose of this line `gpio_init_af(UART_RX_IO, IO_MODE_INPUT_PD, UART_IO_AF, IO_SPEED_VERYFAST);` (https://github.com/olliw42/mLRS/blob/main/mLRS/CommonTx/jr_pin5_interface.h#L134) is NOT to configure the pin as input, but to change from pullup (which the uart library does) to pulldown.

for the swapped case, setting the now Tx pin to pulldown doesn't make any sense and is likely indeed inconsequential. However, IMHO, the now Rx pin should be configured to pulldown. I agree, if it's necessary to also configure the now Tx pin is indeed not so obvious. Ii is probably configured to push pull, but I guess I would not bet on it and hence rather  ensure this.

it's always good to check what CubeMx is doing, and it's in fact how I do all the stm32-ll lib code. And I also noted that it does pushpull for both rx&tx. HOWEVER, CubeMX does have bugs (there is remarkable one e.g. with respect to DMA), and not only one, so we should not take it for granted and let it overrule what we are doing if it doesn't make sense.

the problem with not setting the pull on the rx pin ... which may libraries in fact do, even so called high quality industrial grade libraries! ... is that it can lead to all sort of nonsense when the pin is not connected or not properly connected. I can tell you that over the last more than a decade of using STM32 this has caused more than one issue. It's as always, it may seem to work, and does work most of the time, but if it's not electrical sound it's just a matter of time. So, CubeMX's code will work, but it won't be difficult to create a practical situation where the code will just screw itself up because of e.g. interrupt storm.

I frankly admit that I'm quite surprised that you seem to find that it kind of always works irrespectively of what you do. I remember having spend substantial time with investigating the jrpin5, especially how it is done by others, and working out how to do it properly, and I found that to pay off in terms of reliability. 

Anyway, let's just make it as proper as we can. And let's not recommend to users to use resistors instead of the diode. ;)

Reply:
i.e., I would suggest to have two gpio_init_af after the Enable.

Reply:
Thanks for all the detail here.  Added the two gpio_init_af lines to reflect the new pin types after the uart swap.

Tested successfully with diode using the 's' pad on E77.

Reply:
FANTASTIC!!